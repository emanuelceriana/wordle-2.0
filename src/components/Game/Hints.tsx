import { useContext, useState, useCallback, useMemo } from "react";
import { IAppContext, AppContext } from "../../context/AppContext";
import { useQuery, UseQueryResult } from "react-query";
import { fetchHints, SimplifiedHint } from "../../api";
import styles from "./Hints.module.scss";
import { faCaretLeft, faCaretRight } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import cn from "classnames";

export const Hints = () => {
  const [indexHint, setIndexHint] = useState(0);
  const [showHints, setShowHints] = useState(false);
  const { randomWord, playSound, playlist } =
    useContext<IAppContext>(AppContext);

  const {
    data: hints,
    error,
    refetch,
    isLoading,
  }: UseQueryResult<SimplifiedHint[], Error> = useQuery(
    ["hints", randomWord],
    () => fetchHints(randomWord),
    { enabled: false }
  );

  const handleHints = useCallback(() => {
    refetch();
    !isLoading && playSound(playlist.Reveal);
    setShowHints(true);
  }, [hints, indexHint, isLoading, playSound]);

  const showLeftButton = useMemo(() => {
    return indexHint !== 0;
  }, [hints, indexHint]);

  const showRigthButton = useMemo(() => {
    return hints && hints.length && indexHint < hints.length - 1;
  }, [hints, indexHint]);

  const handleArrowClick = useCallback(
    (number: number) => {
      playSound(playlist.Click);
      setIndexHint((index: number) => index + number);
    },
    [playSound]
  );

  return (
    <div className={styles.hintBox}>
      {showHints && !isLoading ? (
        error || (!hints?.length && !isLoading) ? (
          <div className={styles.error}>
            Ups... you shouldn't be seeing this! ðŸ˜³
          </div>
        ) : (
          <>
            <div
              data-testid="previousHint"
              className={cn(styles.left, {
                [styles.hidden]: !showLeftButton,
              })}
              onClick={() => handleArrowClick(-1)}
            >
              <FontAwesomeIcon icon={faCaretLeft} />
            </div>
            <div data-testid="hintText" className={styles.hintText}>
              <div className={styles.lexicalCategory}>
                <b>{hints?.[indexHint].lexicalCategory}:</b>
              </div>
              <div className={styles.definition}>
                <i>{hints?.[indexHint].shortDefinitions}</i>
              </div>
            </div>

            <div
              data-testid="nextHint"
              className={cn(styles.right, {
                [styles.hidden]: !showRigthButton,
              })}
              onClick={() => handleArrowClick(1)}
            >
              <FontAwesomeIcon icon={faCaretRight} />
            </div>
          </>
        )
      ) : (
        <div className={styles.hidingHint}>
          {isLoading ? (
            <div>Searching hints... ðŸ•µ</div>
          ) : (
            <>
              <button data-testid="hintButton" onClick={handleHints}>
                Hint
              </button>
              <div>
                <b>Note:</b>{" "}
                <i>
                  Some hints make the game 99% simpler so think twice before
                  clicking ðŸ¤”
                </i>
              </div>
            </>
          )}
        </div>
      )}
    </div>
  );
};
