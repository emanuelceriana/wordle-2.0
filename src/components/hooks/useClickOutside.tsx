import React, { useCallback, useEffect } from "react";

interface ClickOutsideProps {
  rowRefs: React.MutableRefObject<HTMLDivElement[]>;
  activeRowIdx: number;
}

const useClickOutside = ({ rowRefs, activeRowIdx }: ClickOutsideProps) => {
  const handleClickOutside = useCallback(
    (event: MouseEvent) => {
      const targetElement = event.target as Node;
      if (!rowRefs?.current[activeRowIdx]?.contains(targetElement)) {
        setTimeout(() => rowRefs?.current[activeRowIdx]?.focus(), 0);
      }
    },
    [rowRefs, activeRowIdx]
  );

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [handleClickOutside]);
};

export default useClickOutside;
