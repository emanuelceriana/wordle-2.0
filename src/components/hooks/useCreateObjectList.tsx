import { useEffect, useState } from "react";
import { v4 as uuidv4 } from "uuid";

interface ObjectList {
  id: string;
}

interface CreateObjectList {
  length: number;
}

export const useCreateObjectList = ({ length }: CreateObjectList) => {
  const [objectList, setObjectList] = useState<ObjectList[]>([]);

  useEffect(() => {
    const objectsArray: ObjectList[] = new Array(length)
      .fill(0)
      .map(() => ({ id: `grid-${uuidv4()}` }));
    setObjectList(objectsArray);
  }, [length]);

  return {
    objectList,
  };
};
