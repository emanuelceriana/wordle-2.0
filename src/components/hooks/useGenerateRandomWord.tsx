import { useCallback } from "react";
import wordsDS from "../../data/wordsDataSet.json";
import { QueryClient } from "react-query";
import { fetchRandomWord } from "../../api";

interface GetRandomWordProps {
  wordLength: number;
  queryClient: QueryClient;
}

export const useGenerateRandomWord = ({
  wordLength,
  queryClient,
}: GetRandomWordProps) => {
  const generate = useCallback(async () => {
    let possibleWords: string[] = [];

    for (const [key, words] of Object.entries(wordsDS)) {
      if (key === wordLength.toString()) {
        possibleWords = words;
      }
    }

    const randomIndex: number = Math.floor(
      Math.random() * possibleWords.length
    );

    try {
      const randomWord = await queryClient.fetchQuery("randomWord", () =>
        fetchRandomWord(wordLength)
      );
      if (randomWord && typeof randomWord === "string" && randomWord != "") {
        return randomWord;
      } else {
        throw new Error("Invalid Random Word");
      }
    } catch {
      return possibleWords[randomIndex];
    }
  }, [wordLength, queryClient]);

  return {
    generate,
  };
};
