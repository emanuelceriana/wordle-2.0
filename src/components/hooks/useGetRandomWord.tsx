import { useEffect, useState } from "react";
import wordsDS from "../../data/wordsDataSet.json";

interface GetRandomWordProps {
  wordLength: number;
}

export const useGetRandomWord = ({ wordLength }: GetRandomWordProps) => {
  const [randomWord, setRandomWord] = useState<string>();

  useEffect(() => {
    let possibleWords: string[] = [];

    for (const [key, words] of Object.entries(wordsDS)) {
      if (key === wordLength.toString()) {
        possibleWords = words;
      }
    }

    const randomIndex: number = Math.floor(
      Math.random() * possibleWords.length
    );
    setRandomWord(possibleWords[randomIndex]);
  }, [wordLength]);

  return {
    randomWord,
  };
};
