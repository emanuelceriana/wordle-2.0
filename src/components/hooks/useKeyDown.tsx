import {
  KeyboardEvent,
  KeyboardEventHandler,
  useCallback,
  useState,
} from "react";
import { ValidationArray, getAmountOfLetters, validateWord } from "../utils";
import KeyboardSound from "../../assets/keyboardSound.wav";
import NextWordSound from "../../assets/nextWordSound.wav";

interface KeyDownProps {
  isRowActive: boolean;
  setFocusedInputIdx: (idx: number) => void;
  setActiveRowIdx: (idx: number) => void;
  activeRowIdx: number;
  inputValues: string[];
  setInputValues: (inputValues: string[]) => void;
  focusedInputIdx: number;
  wordLength: number;
  randomWord: string;
}

const useKeyDown = ({
  isRowActive,
  setFocusedInputIdx,
  setActiveRowIdx,
  activeRowIdx,
  inputValues,
  setInputValues,
  focusedInputIdx,
  wordLength,
  randomWord,
}: KeyDownProps) => {
  const [wordArrayValidation, setWordArrayValidation] = useState<
    ValidationArray[]
  >([]);

  const handleKeyDown: KeyboardEventHandler<HTMLElement> = useCallback(
    (e: KeyboardEvent) => {
      if (isRowActive) {
        const ks = new Audio(KeyboardSound);
        const nw = new Audio(NextWordSound);
        const keyPressed = e.key.toLowerCase();
        if (
          keyPressed === "enter" &&
          getAmountOfLetters(inputValues) === wordLength
        ) {
          setWordArrayValidation(validateWord(randomWord, inputValues));
          setFocusedInputIdx(0);
          setActiveRowIdx(activeRowIdx + 1);
          nw.play();
        } else if (keyPressed === "backspace") {
          if (
            inputValues[focusedInputIdx] &&
            inputValues[focusedInputIdx] !== ""
          ) {
            const updatedValues = [...inputValues];
            updatedValues[focusedInputIdx] = "";
            setInputValues(updatedValues);
            ks.play();
          } else if (
            (!inputValues[focusedInputIdx] ||
              inputValues[focusedInputIdx] === "") &&
            focusedInputIdx > 0
          ) {
            setFocusedInputIdx(focusedInputIdx - 1);
            ks.play();
          }
        } else if (/^[a-z]$/.test(keyPressed)) {
          const updatedValues = [...inputValues];
          updatedValues[focusedInputIdx] = keyPressed;
          setInputValues(updatedValues);
          ks.play();
          if (focusedInputIdx < wordLength - 1) {
            setFocusedInputIdx(focusedInputIdx + 1);
          }
        }
      }
    },
    [isRowActive, inputValues, focusedInputIdx, setActiveRowIdx]
  );

  return {
    handleKeyDown,
    wordArrayValidation,
  };
};

export default useKeyDown;
