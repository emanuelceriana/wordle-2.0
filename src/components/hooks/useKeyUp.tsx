import {
  KeyboardEvent,
  KeyboardEventHandler,
  useCallback,
  useContext,
  useState,
} from "react";

import { AppContext, IAppContext } from "../../context/AppContext";
import { getAmountOfLetters, validateWord, ValidationArray } from "../../utils";

interface KeyUpProps {
  isRowActive: boolean;
  setFocusedInputIdx: (idx: number) => void;
  setActiveRowIdx: (idx: number) => void;
  activeRowIdx: number;
  inputValues: string[];
  setInputValues: (inputValues: string[]) => void;
  focusedInputIdx: number;
  wordLength: number;
  randomWord: string;
}

const hardModeMessageError = (
  <>
    <b>Hard Mode:</b> The word is not a valid word.
  </>
);

const useKeyUp = ({
  isRowActive,
  setFocusedInputIdx,
  setActiveRowIdx,
  activeRowIdx,
  inputValues,
  setInputValues,
  focusedInputIdx,
  wordLength,
  randomWord,
}: KeyUpProps) => {
  const [wordInputValidation, setWordInputValidation] = useState<
    ValidationArray[]
  >([]);
  const {
    playlist,
    userSettings,
    playSound,
    handleGlobalWordValidation,
    validateRealWord,
    setNotification,
  } = useContext<IAppContext>(AppContext);

  const validate = (randomWord: string, inputValues: string[]) => {
    handleGlobalWordValidation(validateWord(randomWord, inputValues));
    setWordInputValidation(validateWord(randomWord, inputValues));
  };

  const checkHardMode = useCallback(async () => {
    let next = true;
    if (userSettings.isHardModeActive) {
      const isValidRealWord = await validateRealWord(inputValues.join(""));
      if (!isValidRealWord) {
        setNotification(hardModeMessageError);
      }
      next = isValidRealWord;
    }
    return next;
  }, [inputValues, userSettings, validateRealWord, setNotification]);

  const handleKeyUp: KeyboardEventHandler<HTMLElement> = useCallback(
    async (e: KeyboardEvent) => {
      if (isRowActive) {
        const keyPressed = e.key.toLowerCase();
        if (
          keyPressed === "enter" &&
          getAmountOfLetters(inputValues) === wordLength
        ) {
          if (await checkHardMode()) {
            validate(randomWord, inputValues);
            setFocusedInputIdx(0);
            setActiveRowIdx(activeRowIdx + 1);
            playSound(playlist.Try);
          }
        } else if (keyPressed === "backspace") {
          if (
            inputValues[focusedInputIdx] &&
            inputValues[focusedInputIdx] !== ""
          ) {
            const updatedValues = [...inputValues];
            updatedValues[focusedInputIdx] = "";
            setInputValues(updatedValues);
          } else if (
            (!inputValues[focusedInputIdx] ||
              inputValues[focusedInputIdx] === "") &&
            focusedInputIdx > 0
          ) {
            setFocusedInputIdx(focusedInputIdx - 1);
          }
          playSound(playlist.Keyboard);
        } else if (/^[a-z]$/.test(keyPressed)) {
          const updatedValues = [...inputValues];
          updatedValues[focusedInputIdx] = keyPressed;
          setInputValues(updatedValues);
          playSound(playlist.Keyboard);
          if (focusedInputIdx < wordLength - 1) {
            setFocusedInputIdx(focusedInputIdx + 1);
          }
        }
      }
    },
    [
      isRowActive,
      inputValues,
      focusedInputIdx,
      userSettings,
      setActiveRowIdx,
      playSound,
      checkHardMode,
    ]
  );

  return {
    handleKeyUp,
    wordInputValidation,
  };
};

export default useKeyUp;
