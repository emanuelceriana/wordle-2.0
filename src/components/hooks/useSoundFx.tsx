import { useCallback } from "react";

import KeyboardSound from "../../assets/keyboardSound.wav";
import LoseSound from "../../assets/loseSound.wav";
import NextWordSound from "../../assets/nextWordSound.wav";
import WinSound from "../../assets/winSound.wav";
import RevealHint from "../../assets/reveal.wav";
import MouseClick from "../../assets/mouseclick.wav";

export enum PlaylistValidKeys {
  Win = "Win",
  Lose = "Lose",
  Keyboard = "Keyboard",
  Try = "Try",
  Reveal = "Reveal",
  Click = "Click",
}

type Playlist = {
  [K in PlaylistValidKeys]: HTMLAudioElement;
};

const playlist: Playlist = {
  [PlaylistValidKeys.Win]: new Audio(WinSound),
  [PlaylistValidKeys.Lose]: new Audio(LoseSound),
  [PlaylistValidKeys.Keyboard]: new Audio(KeyboardSound),
  [PlaylistValidKeys.Try]: new Audio(NextWordSound),
  [PlaylistValidKeys.Reveal]: new Audio(RevealHint),
  [PlaylistValidKeys.Click]: new Audio(MouseClick),
};

interface SoundFxProps {
  isSoundFxActive: boolean;
}

export const useSoundFx = ({ isSoundFxActive }: SoundFxProps) => {
  const play = useCallback(
    (sound: HTMLAudioElement) => {
      if (isSoundFxActive) {
        sound.currentTime = 0;
        sound.play();
      }
    },
    [isSoundFxActive]
  );

  return {
    play,
    playlist,
  };
};
