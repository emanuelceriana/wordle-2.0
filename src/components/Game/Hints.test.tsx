import { fireEvent, act } from "@testing-library/react";
import * as ReactQuery from "react-query";

import { Hints } from "./Hints";
import { playlist } from "../hooks/useSoundFx";
import { renderWithContext } from "../../test-utils";

interface MockValueContext {
  randomWord: string;
  playSound: jest.Mock<any, any>;
  playlist: typeof playlist;
}

describe("<Hints />", () => {
  const mockValueContext: MockValueContext = {
    randomWord: "test",
    playSound: jest.fn(),
    playlist: playlist,
  };

  const mockedRefetch = jest.fn();
  //   jest.mock("react-query");

  const mockedUseQuery = jest.spyOn(ReactQuery, "useQuery");

  const defaultUseQueryMockedValues = {
    data: [],
    error: null,
    isLoading: false,
    refetch: mockedRefetch,
  } as any;

  mockedUseQuery.mockReturnValue(defaultUseQueryMockedValues);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render without errors", () => {
    const { asFragment } = renderWithContext(<Hints />, mockValueContext);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should fecth hints, play sound and show hints", async () => {
    const hint = [{ lexicalCategory: "adj", shortDefinitions: "just a test" }];

    mockedUseQuery.mockReturnValue({
      ...defaultUseQueryMockedValues,
      data: hint,
    });

    const { getByTestId, queryByText } = renderWithContext(
      <Hints />,
      mockValueContext
    );

    const hintButton = getByTestId("hintButton");
    let lexicalCategory = queryByText("adj");
    let shortDefinitions = queryByText("just a test");

    expect(lexicalCategory).toBe(null);
    expect(shortDefinitions).toBe(null);

    act(() => {
      fireEvent.click(hintButton);
      lexicalCategory = queryByText("adj");
      shortDefinitions = queryByText("just a test");
    });

    expect(mockedRefetch).toBeCalledTimes(1);
    expect(mockValueContext.playSound).toBeCalledTimes(1);
    expect(lexicalCategory).toBeDefined();
    expect(shortDefinitions).toBeDefined();
  });

  it("should show next and previous hint when click arrow buttons", () => {
    const hint = [
      { lexicalCategory: "adj", shortDefinitions: "just a test" },
      { lexicalCategory: "noun", shortDefinitions: "next test" },
    ];

    mockedUseQuery.mockReturnValue({
      ...defaultUseQueryMockedValues,
      data: hint,
    });

    const { getByTestId, queryByText } = renderWithContext(
      <Hints />,
      mockValueContext
    );

    const hintButton = getByTestId("hintButton");

    act(() => {
      fireEvent.click(hintButton);
    });

    const nextHintButton = getByTestId("nextHint");

    act(() => {
      fireEvent.click(nextHintButton);
    });

    let lexicalCategory = queryByText("noun");
    let shortDefinitions = queryByText("next test");

    expect(lexicalCategory).toBeDefined();
    expect(shortDefinitions).toBeDefined();

    const previousHintButton = getByTestId("previousHint");

    act(() => {
      fireEvent.click(previousHintButton);
    });

    lexicalCategory = queryByText("adj");
    shortDefinitions = queryByText("just a test");

    expect(lexicalCategory).toBeDefined();
    expect(shortDefinitions).toBeDefined();
  });

  it("should show correct text when is fetching hints", () => {
    mockedUseQuery.mockReturnValue({
      ...defaultUseQueryMockedValues,
      isLoading: true,
    });

    const { queryByText } = renderWithContext(<Hints />, mockValueContext);

    const loadingText = queryByText("Searching hints...");
    expect(loadingText).toBeDefined();
  });

  it("should show correct text when there's an error", () => {
    mockedUseQuery.mockReturnValue({
      ...defaultUseQueryMockedValues,
      isLoading: false,
      error: true,
    });

    const { queryByText } = renderWithContext(<Hints />, mockValueContext);

    const errorText = queryByText("Ups... you shouldn't be seeing this!");
    expect(errorText).toBeDefined();
  });
});
