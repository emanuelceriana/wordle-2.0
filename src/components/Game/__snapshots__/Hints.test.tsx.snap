// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hints /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="hintBox"
  >
    <div
      class="hidingHint"
    >
      <button
        data-testid="hintButton"
      >
        Hint
      </button>
      <div>
        <b>
          Note:
        </b>
         
        <i>
          Some hints make the game 99% simpler so think twice before clicking ðŸ¤”
        </i>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
